### PR TITLE
[Improvement](join) Support nested loop outer join

### PR DESCRIPTION
**Pull request by Gabriel39**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.89 seconds
 load time: 455 seconds
 storage size: 17184096966 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221104074201_clickbench_pr_39961.html

File: be/src/vec/exec/join/vnested_loop_join_node.cpp [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+                            &tmp_block, offset_stack);
+                    mutable_join_block = MutableBlock(std::move(tmp_block));
+                    if (!status.OK()) {
+                        return status;
+                    }
+                }
+
+                if constexpr (set_build_side_flag) {
+                    if (*eos) {
+                        auto& cols = mutable_join_block.mutable_columns();
```
Comment (HappenLee) [inline be/src/vec/exec/join/vnested_loop_join_node.cpp:209]: the work seems just do one time, many be a big, big block?

File: fe/fe-core/src/main/java/org/apache/doris/planner/CrossJoinNode.java [hunk 2]
```diff
@@ -18,85 +18,125 @@
...
+import java.util.Collections;
 import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 /**
  * Cross join between left child and right child.
  */
-public class CrossJoinNode extends PlanNode {
+public class CrossJoinNode extends JoinNodeBase {
```
[the remark this answers is not in the document]
Comment (Gabriel39) [reply fe/fe-core/src/main/java/org/apache/doris/planner/CrossJoinNode.java:54]: Sure. I will do this in next PR

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**